rust-symcrypt/src/ecc/mod.rs

```diff
@@ -0,0 +1,570 @@
+/// [`CurveType`] provides an enum of the curve types that can be used when creating a key(s)
+/// The current curve types supported is `NistP256`, `NistP384`, and `Curve25519`.
+pub enum CurveType {
+    NistP256,
+    NistP384,
+    Curve25519,
+}
+
+#[derive(Debug)]
+// EcKey is a wrapper around symcrypt_sys::PSYMCRYPT_ECKEY.
+// No drop needed on InnerEcKey, drop is handled by EcKey::Drop();
+pub(crate) struct InnerEcKey (symcrypt_sys::PSYMCRYPT_ECKEY);
+
+// InnerEcCurve is a wrapper around symcrypt_sys::PSYMCRYPT_ECURVE.
+pub(crate) struct InnerEcCurve(pub(crate) symcrypt_sys::PSYMCRYPT_ECURVE);
+
+unsafe impl Send for InnerEcCurve {
+    // TODO: Discuss send/sync for rustls
+}
+
+unsafe impl Sync for InnerEcCurve {
+    // TODO: Discuss send/sync for rustls
+}
+
+// Must drop EcCurve after EcKey is dropped, will always be the case since EcCurve is static.
+impl Drop for InnerEcCurve {
+    fn drop(&mut self) {
+        unsafe {
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptEcurveFree(self.0)
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+/// `EcKeyUsage` will indicate if the [`EcKey`] will be used for [`EcKeyUsage::EcDsa`] or [`EcKeyUsage::EcDh`], or [`EcKeyUsage::EcDhAndEcDsa`].
+pub enum EcKeyUsage {
+    /// When using [`EcKeyUsage::EcDsa`] the intended usage of the key will be only for EcDsa sign or verify.
+    EcDsa,
+
+    /// When using [`EcKeyUsage::EcDh`] the intended usage of the key will be only for EcDh.
+    EcDh,
+
+    /// When using [`EcKeyUsage::EcDhAndEcDsa`] the intended usage of the key will be for both EcDsa and EcDh.
+    EcDhAndEcDsa,
+}
+
+// to_symcrypt_flag converts the EcKeyUsage enum to the corresponding SymCrypt flag and is only needed internally.
+impl EcKeyUsage {
+    pub(crate) fn to_symcrypt_flag(&self) -> u32 {
+        match self {
+            EcKeyUsage::EcDsa => symcrypt_sys::SYMCRYPT_FLAG_ECKEY_ECDSA,
+            EcKeyUsage::EcDh => symcrypt_sys::SYMCRYPT_FLAG_ECKEY_ECDH,
+            EcKeyUsage::EcDhAndEcDsa =>  { symcrypt_sys::SYMCRYPT_FLAG_ECKEY_ECDSA | symcrypt_sys::SYMCRYPT_FLAG_ECKEY_ECDH },
+        }
+    }
+}
+
+#[derive(Debug)]
+/// ECC Key State Object.
+/// 
+/// `EcKey` is a struct that represents an Ec Key. It can hold either just an Ec public key, or an Ec key pair.
+/// To check if if the key has a private key attached, you can use [`EcKey::has_private_key()`]
+pub struct EcKey {
+    inner_key: InnerEcKey,
+    curve_type: CurveType,
+    ec_key_usage: EcKeyUsage,
+    has_private_key: bool
+}
+
+unsafe impl Send for EcKey {
+    // TODO: Discuss send/sync for rustls
+}
+
+unsafe impl Sync for EcKey {
+    // TODO: Discuss send/sync for rustls
+}
+
+// Must drop the EcKey before the expanded EcCurve is dropped
+// EcCurve has static lifetime so this will always be the case.
+impl Drop for EcKey {
+    fn drop(&mut self) {
+        unsafe {
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptEckeyFree(self.inner_key());
```

Comment:
>self.inner_key() [](http://example.com/codeflow?start=44&length=16)

If SymCryptEckeyAllocate fails, won't this be null? Might need to check and only free if not null.